### ISSUE TITLE
Indent for an expression spanning multiple lines

### ISSUE THREAD

**Issue by elyzion** [bbatsov/rubocop#1646]
Rubocop complains about the indentation on the .new and .permit_method_call lines in the code snippet below. Aligning the calls with the some_instance line resolves the complaints, but in this case it should be calculating the indent from the SomeService line. I am using version 0.29.0.

I also included my .rubocop.yml below.  Is this a Rubocop limitation, or is it something I can resolve with my configuration? (I tried the indent and align configuration options to no avail.)

```Ruby
some_instance.member.attributes =
  SomeService
    .new(@class_member)
    .permit_method_call(params[:some_key])
```
```Ruby
AlignParameters:
  EnforcedStyle: with_fixed_indentation

AsciiComments:
  Enabled: false

ClassLength:
  Max: 500

CyclomaticComplexity:
  Max: 10

LineLength:
  Max: 140

MethodLength:
  Max: 20

StringLiterals:
  EnforcedStyle: double_quotes
```

**Comment by mvz** [comment]
If I'm not mistaken, it's not the AlignParameters cop that complains here, but instead MultilineOperationIndentation.

**Comment by elyzion** [comment]
Sorry for the long break in communications. Using the following .rubocop.yml still results in the complaints about indentation.
 
```
AbcSize:
  Max: 32

AlignParameters:
  EnforcedStyle: with_fixed_indentation

AsciiComments:
  Enabled: false

ClassLength:
  Max: 500

CyclomaticComplexity:
  Max: 10

LineLength:
  Max: 140

MethodLength:
  Max: 32

MultilineOperationIndentation:
  EnforcedStyle: indented

StringLiterals:
  EnforcedStyle: double_quotes
```
I included the complaints below for reference, I hope this is just a configuration issue on our side!

```
app/services/redacted_service.rb:34:9: C: Use 2 (not 4) spaces for indenting an expression spanning multiple lines.
        .new(@class_member)
        ^^^^
app/services/redacted_service.rb:35:9: C: Use 2 (not 4) spaces for indenting an expression spanning multiple lines.
        .permit_method_call(params[:some_key])
        ^^^^^^^^^^^^^^^^^^^
```

**Comment by jonas054** [comment]
No, you can't solve this with configuration, unless you want to disable the cop. Looking at this issue and at #1633, it seems to me that multiline method call chains should be checked by a separate cop, as people have different expectations and preferences for expressions like the one above compared to things like
```ruby
some_instance.member.attributes =
  some_value +
  new_class_member +
  params[:some_key]
```

**Comment by alexdowad** [comment]
Is this solved with @jonas054's  new `MultilineMethodCallIndentation` cop?

**Comment by jonas054** [comment]
Not exactly. The current `master` supports the following styles. With
```yaml
Style/MultilineMethodCallIndentation:
  EnforcedStyle: indented
```
RuboCop accepts
```ruby
some_instance.member.attributes = SomeService
  .new(@class_member)
  .permit_method_call(params[:some_key])
```
and
```ruby
some_instance.member.attributes =
  SomeService
  .new(@class_member)
  .permit_method_call(params[:some_key])
```
With
```yaml
Style/MultilineMethodCallIndentation:
  EnforcedStyle: aligned
```
it accepts
```ruby
some_instance.member.attributes = SomeService
                                  .new(@class_member)
                                  .permit_method_call(params[:some_key])
```
and

```ruby
some_instance.member.attributes =
  SomeService
  .new(@class_member)
  .permit_method_call(params[:some_key])
```

**Comment by alexdowad** [comment]
@jonas054 OK, thanks. It seems like the style requested here is quite reasonable. Do you have any ideas what a 3rd style for `MultilineMethodCallIndentation` could be called?

**Comment by jonas054** [comment]
Here's my understanding of what is requested.

The style is a variation on the `indented` style, and it differs from the current `indented` style when the expression is a binary operation. In this new style, the base for the indentation is the right hand side of the binary operation rather than the whole expression (or the left hand side).

If my understanding is correct, it follows that the following indentation would be enforced, for example.
```ruby
some_instance.member.attributes == SomeService
                                     .new(@class_member)
                                     .permit_method_call(params[:some_key])
```

The name that springs to mind is `indented_rhs`, but you might be able to come up with a better suggestion, @alexdowad.

**Comment by alexdowad** [comment]
I don't think this style has anything specifically to do with assignments. If the selector in a dotted method call is pushed to a following line, then you follow the chain of dotted method calls back to its root, and indent from the receiver of the first call in the chain.

So the example you gave is right. But so are these:

```ruby
if Something
     .call1
     .call2

while Something
        .call1
        .call2

class A < Something
            .call
```

**Comment by alexdowad** [comment]
@elyzion, does that seem to describe what you are looking for?

**Comment by maia** [comment]
Looking forward to a style that allows:
````
@something = Someclass
                 .where(kind: :hello)
                 .includes(:other_class)
                 .order(created_at: :desc)
````
(using the 4 char standard continued indentation from Rubymine)

**Comment by alexdowad** [comment]
Could we call this style `offset` (from the receiver)? `relative` (to the receiver)? Or give it a long and wordy name which describes it more precisely?

**Comment by jonas054** [comment]
I'd go for long and wordy. How about `indented_from_receiver`, or even `indented_relative_to_receiver`?

**Comment by nicolas-fricke** [comment]
Has there been any progress on this issue? In our project we're now disabling this cop as we're normally writing our code `indented_relative_to_receiver` (I like the name!).

What we'd like to have is something to accept the following code:
```ruby
my_variable = MyClass
                .method_1(123)
                .method_2
```

but keep other multiline statements `aligned`, e.g.:
```ruby
my_variable = if alive?
                42
              else
                0
              end
```

If no one is working on this, I might have a look in how to make this possible myself.

**Comment by dsager** [comment]
Maybe it makes sense to implement a separate cop specifically for method chaining?!

**Comment by aried3r** [comment]
I'm not sure I'm reading the code correctly, but shouldn't it *almost* handle this case already?
I found [this code in the cop](https://github.com/bbatsov/rubocop/blob/f20fdeac6008319e731257bf2bf4e7e27eccefc9/lib/rubocop/cop/style/multiline_method_call_indentation.rb#L100-L103) but can't really make sense of it, because it seems that line either goes unused or ignored (hence this issue).

@nicolas-fricke I'd love to help out, we're also disabling this cop because of this same issue.

**Comment by jonas054** [comment]
Yes, *almost* being the operative word. What you see there is *alignment* where the right hand side of the expression (`b` and `.c`) are aligned with each other without any extra indentation. As I've said earlier in this discussion, we need another style an addition to `aligned` and `indented`. It's doable but not trivial to implement.

**Comment by nicolas-fricke** [comment]
Thanks @jonas054 for taking care of this – looking forward to upgrading rubocop to the next version and then reenabling this new cop :)
